https-bastion.md

```diff
@@ -0,0 +1,138 @@
+# HTTPS Bastion
+
+This document specifies an HTTPS bastion, a service that acts as a reverse proxy
+accepting connections from both clients and backends and routing requests from
+the former to the latter.
+
+Bastions allow backends to run in an environment without a publicly reachable
+address, while still being reachable synchronously over a public HTTP endpoint.
+
+The design requires no changes to HTTPS clients, and can be implemented as an
+abstraction or even a separate service on the backend side with no changes to
+most HTTP servers.
+
+Backends are authenticated by their Ed25519 key, and the hash of that public key
+becomes the first segment of the path of client requests. To prevent abuse, it
+is RECOMMENDED that bastions apply an allowlisting or registration policy to the
+backends they will serve.
+
+Like regular HTTPS reverse proxies, the bastion is in the position of blocking,
+observing, or manipulating the backend’s traffic. Bastions SHOULD only be used
+for self-authenticating protocols, that don't rely exclusively on TLS for
+authentication or confidentiality. Alternatively, backends MAY use bastions they
+trust as part of their broad hosting infrastructure.
+
+Bastions were designed to enable synchronous [transparency log witnesses][]
+which prefer not to accept connections directly from the Internet.
+
+## API
+
+At a high level, backends connect to the bastion authenticating with their
+Ed25519 key and start serving HTTP/2 on that connection. The bastion then
+proxies to that connection all requests that include the backend's public key
+hash in the URL path.
+
+### Backend to bastion connection
+
+A backend connects to the bastion’s backend endpoint with TLS 1.3, specifying
+the ALPN protocol `bastion/0`. Other TLS versions MUST be rejected, as only TLS
+1.3 hides client certificates from network observers. The backend MUST present as
+its client certificate a self-signed Ed25519 certificate containing the backend
+public key. The backend verifies the bastion’s TLS certificate chain like usual.
+The bastion checks the backend public key against an allow list.
+
+After opening the connection, the backend starts serving HTTP/2 traffic on it as
+if it was a client-initiated connection. HTTP/2’s multiplexing allows serving
+multiple parallel requests on a single connection efficiently. None of the
+backend's HTTP APIs are modified, except for observing the `X-Forwarded-For`
+header.
+
+Appendix A presents an example Go adapter that turns a regular HTTP server into
+a bastion backend.
+
+### Client to bastion requests
+
+The bastion accepts HTTP requests at
+
+```
+https://<bastion host>/<key hash>/<path>
+```
+
+where "key hash" MUST be a hex-encoded SHA-256 hash of a 32-byte Ed25519
+public key.
```

Comment:
Consider referencing https://datatracker.ietf.org/doc/html/rfc4648#section-8 for hex, or specify case-insensitive just to be less ambiguous.

Reply:
Good catch. Went for requiring lowercase not to create an infinite set of equivalent URLs. Let me know if you think that's an issue.